chartfx-chart/src/main/java/io/fair_acc/chartfx/axes/spi/format/DefaultFormatter.java

```diff
@@ -77,84 +66,142 @@ public DefaultFormatter(final Axis axis, final String prefix, final String suffi
-                                                               .numberDigitsUntilFirstSignificantDigit(lower, upper);
-                maxSigDigits = Math.max(maxSigDigits, significantDifferentDigits);
-
-                // first tuple index is exp, second is number of fraction digits
-                final Tuple<Double, Double> formatTuple = DigitNumberArithmetic.formatStringForSignificantDigits(lower,
-                        upper);
-                maxExp = (int) Math.max(formatTuple.getXValue(), maxExp);
-                // rstein: added default +1 as temporary mitigation to increase significant number of digits in axis
-                // labels TODO: find a more holistic solution -> May 2019
-                maxFrac = (int) Math.max(formatTuple.getYValue() + 1, maxFrac);
+    void configureFormatter(double range, List<Double> tickMarks) {
+        // Prepare enough cacheable objects
+        final int n = tickMarks.size();
+        while (decompositions.size() < n) {
+            decompositions.add(new Schubfach.DecomposedDouble());
+        }
+
+        // Decompose the double values into significand and exponents
+        for (int i = 0; i < n; i++) {
+            double value = tickMarks.get(i) / unitScaling;
+            if (Math.abs(value) < 1E-14 && range > 1E-12) {
+                // treat rounding errors around zero as zero
+                // TODO:
+                //  (1) confirm that the threshold is sane
+                //  (2) The exp format still renders zero because
+                //      it gets the raw number. This might be easier
+                //      to fix in the tick marker generator?
+                value = 0;
             }
+            Schubfach.decomposeDouble(value, decompositions.get(i));
+        }
+
+        // Special case if we only render a single tick
+        if (n == 1) {
+            var decomp = decompositions.get(0);
+            formatter.setExponentialForm(decomp.getExponent() < -3 || decomp.getExponent() > 4);
+            formatter.setPrecision(-1);
+            return;
+        }
 
-            final StringBuilder sb = new StringBuilder("%");
-            sb.append(maxExp + maxFrac + 1);
-            if (maxFrac > 0) {
-                sb.append('.');
-                sb.append(maxFrac);
-            } else if (maxExp == 0 && maxFrac == 0) {
-                sb.append(".1");
-            } else if (maxExp >= 0 && maxFrac == 0) {
-                sb.append(".0");
+        // Determine the min and max exponents
+        int minExp = decompositions.get(0).getExponent();
+        int maxExp = minExp;
+        for (int i = 1; i < n; i++) {
+            final int exp = decompositions.get(i).getExponent();
+            minExp = Math.min(minExp, exp);
+            maxExp = Math.max(maxExp, exp);
+        }
+
+        // Use the exponential form for very large or very small numbers
+        boolean useExponentialForm = minExp < -3 || maxExp > 4;
```

Comment:
Would it make sense to have these configurable? This would also allow to force exponential form (min/maxExp=0) or decimal form (min/maxExp = MIN/MAX_INT).
Also a third condition could be the difference between min and max exponent: If you have labels with 1.5e-2 and 1.5e+3 its easier to read than 0.15 and 1500.00? I think the way the tickmarks are generated this should only be triggered by log axes and if we make it configurable users can select if they want the before or after behavior from your screenshots.

Reply:
My personal preference is actually for `0.15` and `1500`. I added a `DefaultFormatter::setDisplayFormat(enum)` with `Auto`, `Scientific`, and `Plain` options to force a specific format. I also moved the range check into a protected method for the rare use cases where users may want to override the auto behavior.